useful_resources/sample_code/opt_outs_everyaction.py

```diff
@@ -0,0 +1,163 @@
+import os
+import ast
+import requests
+import time
+from parsons import Redshift, Table, VAN
+from parsons import logger
+from datetime import datetime
+
+# Credentials
+
+# The API_KEYS_STR variable contains API keys for EA committees in this format:
+# {"committee": "Comm. 1 Name", "committee_id": "Comm. 1 ID", "api_key": "Comm. 1 API key"},
+# {"committee": "Comm. 2 Name", "committee_id": "Comm. 2 ID", "api_key": "Comm. 2 API key"}
+API_KEYS_STR = os.environ['VAN_API_KEYS_PASSWORD']
+
+# Configuration Variables
+
+# The OPT_OUT_TABLE is a schema.table of phones to opt out.
+# The table must contain the columns phone, committeeid, and vanid.
+OPT_OUT_TABLE = os.environ['OPT_OUT_TABLE']
+
+# The SUCCESS_TABLE is a schema.table where successful opt-outs will be logged.
+# This table's columns will be: vanid, phone, committeeid, and applied_at.
+SUCCESS_TABLE = os.environ['SUCCESS_TABLE']
+
+# The ERROR_TABLE is a schema.table where errors will be logged.
+# This table's columns will be : vanid, phone, committeeid, errored_at, and error.
+ERROR_TABLE = os.environ['ERROR_TABLE']
+
+# To use the Redshift connector, set the following environmental variables:
+# REDSHIFT_USERNAME
+# REDSHIFT_PASSWORD
+# REDSHIFT_HOST
+# REDSHIFT_DB
+# REDSHIFT_PORT
+
+rs = Redshift()
+
+
+def attempt_optout(ea, row, applied_at, committeeid, success_log, error_log, attempts_left=3):
+
+    vanid = row['vanid']
+    phone = row['phone']
+
+    # Documentation on this json construction is here
+    # https://docs.ngpvan.com/reference/common-models
+    json = {
+        "phones": [
+            {"phoneNumber": phone,
+             "phoneOptInStatus": "O"}
+        ]
+    }
+
+    try:
+        r = ea.update_person_json(id=vanid, match_json=json)
+
+        if isinstance(r, dict):
+            # Response is only a dict upon success
+            success_log.append({
+                "vanid": r.get('vanId'),
+                "phone": phone,
+                "committeeid": committeeid,
+                "applied_at": applied_at
+            })
+
+    # If we get an HTTP Error we log it
+    # Usually these errors are due to a vanid being deleted or merged
+    except requests.exceptions.HTTPError as e:
+        r = e
+        error_log.append({
+            "vanid": vanid,
+            "phone": phone,
+            "committeeid": committeeid,
+            "errored_at": applied_at,
+            "error": r
+        })
+
+    # If we get a connection error we wait a bit and try again.
+    except requests.exceptions.ConnectionError as c:
+        logger.info("Got disconnected, waiting and trying again")
+
+        while attempts_left > 0:
+            attempts_left -= 1
+
+            # Wait 10 seconds, then try again
+            time.sleep(10)
+            attempt_optout(ea, row, attempts_left)
+
+        else:
+            r = str(c)[:999]
```

Comment:
What's happening here? Are connection errors much longer than HTTP errors? Why is this being reformatted when the other one wasn't?